src/engine.rs

```diff
@@ -288,6 +344,58 @@ impl Engine {
         sdk::sha256(&data)
     }
 
+    pub fn charge_gas_limit(
+        sender: &Address,
+        gas_limit: U256,
+        gas_price: U256,
+    ) -> Result<Wei, GasPaymentError> {
+        // Early exit as performance optimization
+        if gas_price.is_zero() {
+            return Ok(Wei::zero());
+        }
+
+        let eth_amount = Wei::new(
+            gas_limit
+                .checked_mul(gas_price)
+                .ok_or(GasPaymentError::EthAmountOverflow)?,
+        );
+        let account_balance = Self::get_balance(sender);
+        let remaining_balance = account_balance
+            .checked_sub(eth_amount)
+            .ok_or(GasPaymentError::OutOfFund)?;
```

Comment:
```suggestion
            .ok_or(GasPaymentError::OutOfFunds)?;
```

Reply:
This one is fine, imo. In upstream EVM its also `OutOfFund`

Reply:
Yes, I was trying to be consistent with the SputnikVM error.

Reply:
Yes, I know that, but `OutOfFund` is incorrect, I suggest using the correct English version instead.